### PR TITLE
update to use tagged unions

### PR DESCRIPTION
**Pull request by pchickey**
DO NOT MERGE until upstream WASI changes are merged into master, and submodule is pointed at master.

Implements https://github.com/WebAssembly/WASI/pull/220/

### CODE REVIEW

**Comment by kubkon** [comment]
@pchickey I've now submitted #42 that fast-forwards `WASI` to the latest master, and fixes a lot of the issues you'd have to fix in this PR that are not directly relevant to the tagged unions proposal. So hopefully this will be useful to you!

**Comment by pchickey** [comment]
@kubkon many thanks, ive rebased

File: src/lib_generated.rs [hunk 1]
```diff
@@ -6,7 +6,7 @@ use core::mem::MaybeUninit;
 
 pub use crate::error::Error;
 pub type Result<T, E = Error> = core::result::Result<T, E>;
-pub type Size = usize;
+pub type Size = u32;
```
Comment (alexcrichton) [inline src/lib_generated.rs]: Was this intentional? IIRC this was a convenience from before because `usize` is used so ubiquitously in Rust right nwo

**Comment by pchickey** [comment]
Oops, that was not intentional, I'll revert it. We now have BuiltinType::Size in witx but the snapshot isnt using it yet, so I'll leave the hack to make the target of the type named `"size"` a `usize` rather than `u32`.

**Comment by alexcrichton** [comment]
👍

**Pull request merged**